### PR TITLE
Kademlia DHT implementation in py-libp2p

### LINKED ISSUE #540: Add Support for libp2p-kad-dht in py-libp2p
### Description

This feature request proposes the integration of the libp2p-kad-dht module into py-libp2p, enabling decentralized key-value storage and lookup. Adding Kademlia DHT capabilities will align py-libp2p with other implementations such as js-libp2p and go-libp2p, enhancing interoperability and use cases like peer discovery and distributed content addressing.

### Motivation

- Decentralized Key-Value Store: Provides a scalable and efficient mechanism for storing and retrieving data.
 
- Improved Peer Discovery: Enables nodes to discover peers dynamically in distributed environments.
 
- Parity Across Implementations: Bridges the gap with js-libp2p and go-libp2p, making py-libp2p more feature-complete.

- Expanding Use Cases: Facilitates applications in file sharing, decentralized storage, and more.

### Current Implementation

- Integration: Implement the libp2p-kad-dht module in py-libp2p.

- Routing Table: Add a RoutingTable implementation for efficient peer and value lookup.

- Protocols: Support for the Kademlia protocols (e.g., FindNode, PutValue, GetValue).

- Bootstrapping: Enable bootstrapping to known peers for initializing the DHT network.

- Testing: Include unit tests and integration tests for verifying DHT operations (e.g., put/get).

Additional Context

Reference Implementations:

- [go-libp2p-kad-dht](https://github.com/libp2p/go-libp2p-kad-dht) 
- [js-libp2p-kad-dht](https://github.com/libp2p/js-libp2p-kad-dht)

Potential Applications: Peer discovery, content routing in IPFS, and decentralized name resolution.

Community Need: Adds significant value to the Python ecosystem, especially for projects building decentralized applications and protocols.

Alternatives Considered: 
Using other implementations (e.g., go-libp2p via bindings), but this adds complexity and deviates from native Python goals.

Expected Challenges
- Maintaining compatibility with the existing py-libp2p architecture.
- Ensuring performance and scalability similar to go-libp2p and js-libp2p.

Resources
- [Kademlia DHT paper](https://pdos.csail.mit.edu/~petar/papers/maymounkov-kademlia-lncs.pdf)
- [libp2p Specification: DHT Protocols](https://github.com/libp2p/specs/tree/master/kad-dht)

### Are you planning to do it yourself in a pull request ?

Maybe

### PR DESCRIPTION
**Pull request by sumanjeet0012**
# Kademlia DHT implementation

## What was wrong?

Kademlia DHT implementation is not present in py-libp2p.
Fixes #540 
## How was it fixed?

Added kademlia DHT module  in py-libp2p

Summary of approach.

Currently features added are:
1. Peer Discovery
2. PUT_VALUE 
3. GET_VALUE
4. ADD_PROVIDERS
5. GET_PROVIDERS

### To-Do

- [ ] Clean up commit history
- [ ] Add or update documentation related to these changes
- [ ] Add entry to the [release notes](https://github.com/libp2p/py-libp2p/blob/main/newsfragments/README.md)

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

**Comment by seetadev** [comment]
@sumanjeet0012 : Neat progress. 

Great work pushing this forward — it’s shaping up really well! 🎉

It’s good to hear that most of the major issues are fixed and that basic documentation is already in place. 

Tagging @pacrob : Hi Paul. Wish if you could also review the PR. Would love to have your inputs too.

On the same note, wish to also share that I did have a discussion with Dave (@dhuseby) on features like DHT client/consumer node creation in Kademlia, which can be tricky, especially around the nuances of routing table behaviors and node modes. He recommended us to refer to https://github.com/libp2p/rust-libp2p/tree/master/examples/ipfs-kad . Would recommend you to study it for reference and also for ensuring compatibility with a key libp2p module.

@acul71 : Kad + identify + yamux + autonat are key building blocks for P2P dapps. Would like your feedback on this PR too.

**Comment by acul71** [comment]
@sumanjeet0012 
I've fixed some doc issues. Can you cherry pick this: https://github.com/libp2p/py-libp2p/pull/617/commits/ab3c88c633c372bf3f29835ff72dbec49902d20d

**Comment by acul71** [comment]
> 
> @acul71 : Kad + identify + yamux + autonat are key building blocks for P2P dapps. Would like your feedback on this PR too.

# Kademlia DHT Implementation Feedback

## Overview
The implementation provides a solid foundation for Kademlia DHT in py-libp2p, maintaining good compatibility with the Rust implementation. This document provides detailed feedback and suggestions for enhancement.

## 1. Node Modes

### Current Implementation
- Basic DHT node implementation without explicit mode distinction
- All nodes currently operate in full node mode

### Suggested Enhancements
1. **Client Mode**
   - Implement lightweight client mode for resource-constrained environments
   - Skip routing table maintenance
   - Only perform essential DHT operations
   - Example configuration:
     ```python
     class KadDHT(Service):
         def __init__(self, host: IHost, mode: str = "full"):
             self.mode = mode  # "full" or "client"
             if mode == "client":
                 self.skip_routing_table_maintenance = True
     ```

2. **Server Mode**
   - Add dedicated server mode for high-performance nodes
   - Enhanced routing table maintenance
   - Priority handling of DHT requests
   - Example:
     ```python
     class KadDHT(Service):
         async def run(self):
             if self.mode == "server":
                 await self._enhanced_routing_maintenance()
             else:
                 await self._standard_routing_maintenance()
     ```

3. **Mode Transitions**
   - Add ability to switch modes at runtime
   - Handle state transitions gracefully
   - Example:
     ```python
     async def switch_mode(self, new_mode: str):
         if new_mode == "client":
             await self._cleanup_routing_table()
         elif new_mode == "server":
             await self._initialize_enhanced_features()
     ```

## 2. Query Optimization

### Current Implementation
- Basic iterative lookup
- Simple peer selection

### Suggested Enhancements
1. **Parallel Query Optimization**
   - Implement parallel query execution
   - Better handling of slow peers
   - Example:
     ```python
     async def find_closest_peers_network(self, target_key: bytes, count: int = 20):
         async with trio.open_nursery() as nursery:
             for peer in peers_to_query:
                 nursery.start_soon(self._query_peer_for_closest, peer, target_key)
     ```

2. **Caching Layer**
   - Add caching for frequent queries
   - Implement TTL-based cache invalidation
   - Example:
     ```python
     class QueryCache:
         def __init__(self, ttl: int = 300):
             self.cache = {}
             self.ttl = ttl
     ```

3. **Query Prioritization**
   - Prioritize queries based on importance
   - Handle timeouts more gracefully
   - Example:
     ```python
     class QueryPriority:
         HIGH = 0
         MEDIUM = 1
         LOW = 2
     ```

## 3. Error Handling

### Current Implementation
- Basic error handling
- Limited error types

### Suggested Enhancements
1. **Custom Error Types**
   - Define specific error types for DHT operations
   - Better error categorization
   - Example:
     ```python
     class DHTError(Exception):
         pass

     class RoutingError(DHTError):
         pass

     class QueryError(DHTError):
         pass
     ```

2. **Error Recovery**
   - Implement automatic recovery mechanisms
   - Handle network partitions gracefully
   - Example:
     ```python
     async def handle_query_error(self, error: DHTError):
         if isinstance(error, RoutingError):
             await self._rebuild_routing_table()
         elif isinstance(error, QueryError):
             await self._retry_with_backoff()
     ```

3. **Error Reporting**
   - Add detailed error reporting
   - Include context in error messages
   - Example:
     ```python
     class ErrorContext:
         def __init__(self, operation: str, peer_id: ID):
             self.operation = operation
             self.peer_id = peer_id
             self.timestamp = time.time()
     ```

## 4. Documentation

### Current Implementation
- Basic API documentation
- Limited examples

### Suggested Enhancements
1. **API Documentation**
   - Add comprehensive API documentation
   - Include usage examples
   - Example structure:
     ```markdown
     # KadDHT
     ## Overview
     The KadDHT class provides...
     
     ## Usage
     ```python
     # Example code
     ```
     ```

2. **Configuration Guide**
   - Document all configuration options
   - Provide tuning guidelines
   - Example:
     ```markdown
     # Configuration
     ## Node Modes
     - Client mode: ...
     - Server mode: ...
     
     ## Performance Tuning
     - Routing table size: ...
     - Query timeout: ...
     ```

3. **Integration Examples**
   - Add examples with other libp2p components
   - Show common use cases
   - Example:
     ```python
     # Example with identify
     async def setup_dht_with_identify():
         host = await create_host()
         identify = Identify(host)
         dht = KadDHT(host)
         await identify.start()
         await dht.start()
     ```

## 5. Testing

### Current Implementation
- Basic unit tests
- Limited integration tests

### Suggested Enhancements
1. **Unit Tests**
   - Add comprehensive unit tests
   - Cover edge cases
   - Example:
     ```python
     def test_routing_table_edge_cases():
         # Test empty routing table
         # Test full routing table
         # Test peer replacement
     ```

2. **Integration Tests**
   - Add integration tests with Rust implementation
   - Test cross-implementation compatibility
   - Example:
     ```python
     async def test_cross_implementation():
         # Setup Python DHT
         # Setup Rust DHT
         # Test communication
     ```

3. **Performance Tests**
   - Add performance benchmarks
   - Compare with Rust implementation
   - Example:
     ```python
     def benchmark_peer_discovery():
         # Measure discovery time
         # Compare with Rust
     ```

## Conclusion
The current implementation provides a solid foundation. Implementing these enhancements would make the Python implementation more robust and feature-complete, while maintaining compatibility with the Rust version.

## References

### Core Implementation
- [Rust libp2p Kademlia Implementation](https://github.com/libp2p/rust-libp2p/tree/master/examples/ipfs-kad)
  - [Kademlia DHT Module](https://github.com/libp2p/rust-libp2p/blob/master/protocols/kad/src/behaviour.rs)
    ```rust
    pub struct KadDHT {
        config: KadConfig,
        mode: Mode,
        routing_table: RoutingTable,
        // ...
    }

    impl KadDHT {
        pub fn new(config: KadConfig) -> Self {
            // Initialize DHT with configuration
        }
    }
    ```
  - [Routing Table Implementation](https://github.com/libp2p/rust-libp2p/blob/master/protocols/kad/src/kbucket.rs)
    ```rust
    pub struct KBucket {
        peers: Vec<Peer>,
        last_updated: Instant,
        // ...
    }

    impl KBucket {
        pub fn insert(&mut self, peer: Peer) -> bool {
            // Insert peer if bucket not full
            // or replace least recently seen peer
        }
    }
    ```
  - [Query Implementation](https://github.com/libp2p/rust-libp2p/blob/master/protocols/kad/src/query.rs)
    ```rust
    pub struct Query {
        target: Key,
        peers: Vec<PeerId>,
        // ...
    }

    impl Query {
        pub async fn run(&mut self) -> QueryResult {
            // Execute parallel queries
            // Handle timeouts and errors
        }
    }
    ```

### Node Modes
- [Client Mode Implementation](https://github.com/libp2p/rust-libp2p/blob/master/protocols/kad/src/behaviour.rs#L100)
  ```rust
  pub enum Mode {
      Client,
      Server,
  }

  impl KadDHT {
      pub fn as_client() -> Self {
          KadDHT {
              mode: Mode::Client,
              // Skip routing table maintenance
              // Only perform essential operations
          }
      }
  }
  ```
- [Server Mode Configuration](https://github.com/libp2p/rust-libp2p/blob/master/protocols/kad/src/behaviour.rs#L200)
  ```rust
  impl KadDHT {
      pub fn as_server() -> Self {
          KadDHT {
              mode: Mode::Server,
              // Enhanced routing table maintenance
              // Priority request handling
          }
      }
  }
  ```

### Query Optimization
- [Parallel Query Implementation](https://github.com/libp2p/rust-libp2p/blob/master/protocols/kad/src/query.rs#L150)
  ```rust
  impl Query {
      async fn parallel_query(&mut self) -> Vec<PeerId> {
          let mut handles = vec![];
          for peer in self.peers.iter() {
              handles.push(self.query_peer(peer));
          }
          futures::future::join_all(handles).await
      }
  }
  ```
- [Query Timeout Handling](https://github.com/libp2p/rust-libp2p/blob/master/protocols/kad/src/query.rs#L300)
  ```rust
  impl Query {
      async fn query_with_timeout(&mut self, peer: &PeerId) -> Result<(), Error> {
          tokio::time::timeout(
              Duration::from_secs(5),
              self.query_peer(peer)
          ).await.map_err(|_| Error::Timeout)
      }
  }
  ```

### Error Handling
- [Error Types](https://github.com/libp2p/rust-libp2p/blob/master/protocols/kad/src/error.rs)
  ```rust
  #[derive(Debug, Error)]
  pub enum Error {
      #[error("Query timed out")]
      Timeout,
      #[error("Routing error: {0}")]
      Routing(String),
      #[error("Query error: {0}")]
      Query(String),
  }
  ```
- [Error Recovery](https://github.com/libp2p/rust-libp2p/blob/master/protocols/kad/src/behaviour.rs#L400)
  ```rust
  impl KadDHT {
      async fn handle_error(&mut self, error: Error) {
          match error {
              Error::Timeout => self.retry_with_backoff().await,
              Error::Routing(_) => self.rebuild_routing_table().await,
              Error::Query(_) => self.cleanup_failed_queries().await,
          }
      }
  }
  ```

### Testing
- [Integration Tests](https://github.com/libp2p/rust-libp2p/blob/master/protocols/kad/src/tests.rs)
  ```rust
  #[tokio::test]
  async fn test_dht_operations() {
      let mut dht1 = KadDHT::new(config1);
      let mut dht2 = KadDHT::new(config2);
      
      // Test peer discovery
      let peer = dht1.find_peer(dht2.local_peer_id()).await;
      assert!(peer.is_some());
  }
  ```
- [Performance Tests](https://github.com/libp2p/rust-libp2p/blob/master/protocols/kad/src/tests.rs#L200)
  ```rust
  #[bench]
  fn bench_peer_discovery(b: &mut Bencher) {
      let mut dht = KadDHT::new(config);
      b.iter(|| {
          dht.find_peer(target_peer_id)
      });
  }
  ```

### Specifications and Papers
- [Kademlia Paper](https://pdos.csail.mit.edu/~petar/papers/maymounkov-kademlia-lncs.pdf)
- [libp2p DHT Specification](https://github.com/libp2p/specs/tree/master/kad-dht)

### Examples
- [Basic DHT Example](https://github.com/libp2p/rust-libp2p/blob/master/examples/ipfs-kad/src/main.rs)
  ```rust
  #[tokio::main]
  async fn main() {
      let mut dht = KadDHT::new(config);
      dht.bootstrap(&bootstrap_peers).await;
      
      // Start DHT operations
      dht.start().await;
  }
  ```
- [Client Mode Example](https://github.com/libp2p/rust-libp2p/blob/master/examples/ipfs-kad/src/main.rs#L100)
  ```rust
  let dht = KadDHT::as_client();
  // Client mode only performs essential operations
  dht.find_peer(target_peer_id).await;
  ```
- [Server Mode Example](https://github.com/libp2p/rust-libp2p/blob/master/examples/ipfs-kad/src/main.rs#L200)
  ```rust
  let dht = KadDHT::as_server();
  // Server mode maintains routing table and handles requests
  dht.start().await;
  ```

**Comment by acul71** [comment]
@sumanjeet0012 
Hi Sumanjeet, I can see that there are are a couple of issues:
- [docs/readthedocs.org:py-libp2p](https://app.readthedocs.org/projects/py-libp2p/builds/28359099/) — Read the Docs build failed!
  # ReadTheDocs Build Error: Missing `kademlia_pb2.py`

## Error Message

During the ReadTheDocs build, the following errors are reported:

```
WARNING: autodoc: failed to import module 'kad_dht.kad_dht' from module 'libp2p'; the following exception was raised:
No module named 'libp2p.kad_dht.pb.kademlia_pb2' [autodoc.import_object]
...
```

## Cause

This error occurs because the file `libp2p/kad_dht/pb/kademlia_pb2.py` is **not present** in the GitHub repository. While it may exist locally (generated by `protoc`), it is either:

- Not committed to the repository, or
- Not generated during the ReadTheDocs build process.

ReadTheDocs builds your documentation from the files in your GitHub repository. If the generated file is missing, imports will fail.

## Solution

You have two options:

### 1. Commit the Generated File

1. Add and commit the file:
   ```sh
   git add libp2p/kad_dht/pb/kademlia_pb2.py
   git commit -m "Add generated kademlia_pb2.py"
   git push
   ```
2. Trigger a rebuild on ReadTheDocs.

### 2. Generate the File During the RTD Build

Update your `.readthedocs.yaml` to generate the file during the build.  

A similar error happened to me in the past in another PR. 
In that PR I asked Paul if he wanted me to `git add and commit` the `pb2` files but he answered me the he would take care of the issue so ping @pacrob for suggestions (I think he made the file generate in some way)

**Comment by acul71** [comment]
@sumanjeet0012 
the other issue:

- [ProviderRecord Boundary Condition Test Failures (Windows)](https://github.com/libp2p/py-libp2p/actions/runs/15367912558/job/43243186898?pr=579)

This is failing only in `windows`. If you can setup a (virtual?) windows dev enviroment and experimenting there, maybe adding logging info, and try to solve the issue. I did it once for another issue, but sometimes it wasn't failing, so it was difficult to debug.

If you can't, try submit a solution and see if test passed in github. 

Here's an **GitHub Codespaces** with **GitHub Copilot** AI integration suggested solution even thought I don't have much faith in it, and I don't like it very much. I would prefere to see some debug logging to see what's happening.

# ProviderRecord Boundary Condition Test Failures (Windows)

## Problem

Two tests are failing on Windows due to boundary condition logic in `ProviderRecord`:

- `test_is_expired_boundary_condition`
- `test_should_republish_boundary_condition`

**Error:**  
At the exact boundary (`timestamp == now - INTERVAL`), the tests expect the record to be expired or need republishing, but the implementation uses `>` (greater than), so the result is `False`.

---

## Cause

The implementation in `ProviderRecord` likely looks like:

```python
def is_expired(self):
    return time.time() - self.timestamp > PROVIDER_RECORD_EXPIRATION_INTERVAL

def should_republish(self):
    return time.time() - self.timestamp > PROVIDER_RECORD_REPUBLISH_INTERVAL
```

At the boundary, `time.time() - self.timestamp == INTERVAL`, so `>` is `False`.

---

## Why does this only fail on Windows?

This test passes on Linux but fails on Windows likely due to differences in how time is measured and floating-point precision between the two operating systems.  
On Linux, `time.time() - boundary_timestamp` might be **slightly greater than** `INTERVAL` due to timing or floating-point rounding, so `>` returns `True` and the test passes.  
On Windows, the calculation might be **exactly equal** to `INTERVAL` (or even slightly less), so `>` returns `False` and the test fails.  
Boundary tests are sensitive to these tiny differences.

---

## Solution

### Recommended: Update the Implementation

Change the comparison from `>` to `>=` so that the record is considered expired or needs republishing **at** the boundary:

```python
def is_expired(self):
    return time.time() - self.timestamp >= PROVIDER_RECORD_EXPIRATION_INTERVAL

def should_republish(self):
    return time.time() - self.timestamp >= PROVIDER_RECORD_REPUBLISH_INTERVAL
```

This will make the tests pass and is the typical expectation for interval-based expiration.

**Will this fix work?**  
Yes, updating the implementation to use `>=` will ensure that the record is considered expired or needs republishing at the exact boundary, making the tests pass consistently on both Linux and Windows. This approach is robust against small timing and floating-point differences between platforms.

---

### Alternative: Update the Tests

If you want the record to be expired or need republishing **after** the interval (not at the boundary), update the tests to expect `False` at the boundary.

---

## Summary

- **Preferred:** Update the implementation to use `>=` for boundary checks.
- **Alternative:** Update the tests to expect `False` at the boundary.

Most systems use `>=` for expiration logic, so updating the implementation is recommended for cross-platform consistency.

**Comment by seetadev** [comment]
@sumanjeet0012 : Great progress. Kad-DHT is coming along very nicely.

**Comment by sumanjeet0012** [comment]
@seetadev The Kademlia DHT is in a Good state now. I believe it’s a good time to proceed with a detailed review from @pacrob.

**Comment by seetadev** [comment]
> @seetadev The Kademlia DHT is in a Good state now. I believe it’s a good time to proceed with a detailed review from @pacrob.

@sumanjeet0012  : Thanks for your wonderful efforts. The CI/CD issues have been resolved indeed. Great work.

We are very near to having kad-dht in production. Appreciate the great initiative. Your steady progress and attention to detail really paid off. Great progress.

As suggested for core py-libp2pmodules, please add user and developer documentation in fine details with screenshots and 2 demo videos/ screencasts (one for users and the other for developers planning to extend and develop using kad-dht). These are modules, which will are critical for the adoption of py-libp2p and also enable the development of multiple initiatives including universal connectivity dapp.

Wish to also thank @acul71 for his continued support and feedback on key issues. Thank you Luca :)


@pacrob The kad-dht module PR in py-libp2p is very near to production.
Thank you so much for your continued support, thoughtful feedback, and valuable guidance — appreciate it. If you have any final thoughts or suggestions, happy to work with @sumanjeet0012 and @acul71  to make sure everything lands in the best possible shape.

**Comment by seetadev** [comment]
@sumanjeet0012 : Great work on the PR. Appreciate your contributions and the quality of the work so far.

When you get a chance, please rebase your branch to align with the latest changes following the merge of PR#618 (please visit https://github.com/libp2p/py-libp2p/pull/618). You may also want to take a look at the related discussion here: https://github.com/libp2p/py-libp2p/discussions/610 for additional context.

For onboarding and setup updates, I’d also recommend checking out the updated guide here: https://github.com/libp2p/py-libp2p/discussions/625. As part of the recent revamp, we're deprecating setup.py in favor of the more modern pyproject.toml, which allows for easier integration with dependency tools like pip, uv, and poetry.

File: examples/kademlia/basic_dht.py [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+
+    """
+    # Get the SHA-256 hash
+    content_hash = hashlib.sha256(content).digest()
+    # For simplicity, we use the raw hash as the key
+    return content_hash
+
+
+async def run_node(
+    port: int, mode: Optional[str], bootstrap_addrs: Optional[list[str]] = None
```
Comment (pacrob) [inline examples/kademlia/basic_dht.py]: It looks like `mode` is unused, and `bootstrap_addrs` is always immediately overwritten.

File: examples/kademlia/basic_dht.py [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+                # val_data = msg.encode()
+                # logger.info(
+                #     f"Storing value with key: {base58.b58encode(val_key).decode()}"
+                # )
+                # await dht.put_value(val_key, val_data)
+                # logger.info(
+                #     f"Stored value with key: {base58.b58encode(val_key).decode()}"
+                # )
+                # logger.info("Value stored is %s", val_data.decode())
+                # trio.sleep(0.5)
```
Comment (pacrob) [inline examples/kademlia/basic_dht.py]: Should this be uncommented or removed?

File: examples/kademlia/basic_dht.py [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+                # Advertise that we can serve this content
+                # logger.info(
+                #     "Advertising as server for content:" f" {content_key.hex()}"
+                # )
+                # success = await dht.provider_store.provide(content_key)
+                # if success:
+                #     logger.info("Successfully advertised as content server")
+                # else:
+                #     logger.warning("Failed to advertise as content server")
+
```
Comment (pacrob) [inline examples/kademlia/basic_dht.py]: Should this be uncommented or removed?

File: examples/kademlia/basic_dht.py [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+                    )
+                else:
+                    logger.warning(
+                        "No servers found for our content %s", content_key.hex()
+                    )
+                # Print bootstrap command for client nodes
+                bootstrap_cmd = f"--bootstrap {addr_str}"
+                logger.info("To connect to this node, use: %s", bootstrap_cmd)
+                print(f"\nTo connect to this node, use: {bootstrap_cmd}\n")
+                # print(f"Content ID for discovery: {content_key.hex()}")
```
Comment (pacrob) [inline examples/kademlia/basic_dht.py]: Should be removed or moved to logging?

**Comment by sumanjeet0012** [comment]
@pacrob Thank you for the feedback.

I’ve made some adjustments to align with the post-revamp setup. However, running `pyrefly typecheck` locally is producing errors with logs spanning over 22,000 lines, so resolving all the issues will take some time.

In the meantime, could you please review the core implementation of the Kademlia DHT?

I’ve also created the Kademlia DHT example file primarily to test whether all functionalities are working as expected. I would appreciate your suggestions on what should be included in the final Kademlia DHT example file.
Would it be appropriate to demonstrate the functionality by adding data to the network? should I include PUT_VALUE and ADD_PROVIDER in the example?

Additionally, I’ve observed that other nodes are not responding to PUT_VALUE and ADD_PROVIDER requests. Could this be because they only accept data from verified nodes or specific addresses?

**Comment by sumanjeet0012** [comment]
@pacrob Ready for review.

File: libp2p/kad_dht/kad_dht.py [hunk 5]
```diff
@@ -0,0 +1,545 @@
...
+        :param mode: The mode of host (Client or Server)
+        """
+        super().__init__()
+
+        self.host = host
+        self.local_peer_id = host.get_id()
+
+        self.mode = mode.upper()
+        if self.mode not in ["CLIENT", "SERVER"]:
+            raise ValueError(f"Invalid mode '{mode}'. Must be 'CLIENT' or 'SERVER'.")
```
Comment (pacrob) [inline libp2p/kad_dht/kad_dht.py]: Consider using an `enum` for mode instead of `str`.

File: libp2p/kad_dht/kad_dht.py [hunk 6]
```diff
@@ -0,0 +1,545 @@
...
+                        response.record.timeReceived = str(time.time())
+
+                        # Serialize and send response
+                        response_bytes = response.SerializeToString()
+                        await stream.write(varint.encode(len(response_bytes)))
+                        await stream.write(response_bytes)
+                        logger.debug("Sent GET_VALUE response")
+                    else:
+                        logger.debug(f"No value found for key {key.hex()}")
+
```
Comment (pacrob) [inline libp2p/kad_dht/kad_dht.py:371]: Should there really be no response if the value is not found locally?

File: libp2p/kad_dht/kad_dht.py [hunk 7]
```diff
@@ -0,0 +1,545 @@
...
+                        response.type = Message.MessageType.PUT_VALUE
+                        response.key = key
+
+                        response_bytes = response.SerializeToString()
+                        await stream.write(varint.encode(len(response_bytes)))
+                        await stream.write(response_bytes)
+                        logger.debug("Sent PUT_VALUE acknowledgement")
+                    else:
+                        logger.warning("Invalid PUT_VALUE message format")
+
```
Comment (pacrob) [inline libp2p/kad_dht/kad_dht.py]: Again, it seems like there should be something raised or other message returned to the other node or user if it fails, not just logging. Please excuse my question if this is not necessary or handled somewhere else.

File: libp2p/kad_dht/utils.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    """
+    # Ensure keys are the same length by using minimum length
+    length = min(len(key_one), len(key_two))
+    result = 0
+
+    # Calculate XOR distance
+    for i in range(length):
+        result = (result << 8) | (key_one[i] ^ key_two[i])
+
+    return result
```
Comment (pacrob) [inline libp2p/kad_dht/utils.py]: This seems like it will truncate the end of the longer key. Should they not be either: 1) required to be the same length, or 2) pad the shorter key in order to not lose information?

File: libp2p/kad_dht/kad_dht.py (hunk 7 above)
Comment (sumanjeet0012) [reply libp2p/kad_dht/kad_dht.py]: Completely agree on your suggestion, It is currently missing, I will implement it as soon as possible.

File: libp2p/kad_dht/kad_dht.py (hunk 5 above)
Comment (sumanjeet0012) [reply libp2p/kad_dht/kad_dht.py]: Changed from string to enum.

File: libp2p/kad_dht/kad_dht.py [hunk 9]
```diff
@@ -0,0 +1,562 @@
...
+    async def put_value(self, key: bytes, value: bytes) -> None:
+        """
+        Store a value in the DHT.
+        """
+        logger.debug(f"Storing value for key {key.hex()}")
+
+        # 1. Store locally first
+        self.value_store.put(key, value)
+        logger.debug(f"Stored value locally for key {key.hex()}")
```
Comment (pacrob) [inline libp2p/kad_dht/kad_dht.py]: nit: it would be nice to log both key and value in at least one of these places

File: libp2p/kad_dht/routing_table.py [hunk 10]
```diff
@@ -0,0 +1,620 @@
...
+        :param count: Maximum number of peers to return
+
+        Returns
+        -------
+            List[ID]: List of peer IDs closest to the key
+
+        """
+        # Get all peers from all buckets
+        all_peers = []
+        # logger.info("bucket is ",self.buckets.)
```
Comment (pacrob) [inline libp2p/kad_dht/routing_table.py]: should be uncommented or removed?

File: libp2p/kad_dht/routing_table.py [hunk 11]
```diff
@@ -0,0 +1,620 @@
...
+        -------
+            int: XOR distance between the keys
+
+        """
+        # Convert to integers
+        k1 = int.from_bytes(key1, byteorder="big")
+        k2 = int.from_bytes(key2, byteorder="big")
+
+        # Calculate XOR distance
+        return k1 ^ k2
```
Comment (pacrob) [inline libp2p/kad_dht/routing_table.py]: We have 2 locations that calculate the distance between peers (here and in `utils.py`). What do you think about consolidating that to one implementation?

File: libp2p/kad_dht/value_store.py [hunk 12]
```diff
@@ -0,0 +1,384 @@
...
+        self.host = host
+        self.local_peer_id = local_peer_id
+
+    def put(self, key: bytes, value: bytes, validity: float = 0.0) -> None:
+        """
+        Store a value in the DHT.
+
+        :param key: The key to store the value under
+        :param value: The value to store
+        :param ttl: Time to live in seconds, or None for no expiration
```
Comment (pacrob) [inline libp2p/kad_dht/value_store.py]: Last arg is `validity` in the definition, but `ttl` in the docstring.

File: pyproject.toml [hunk 13]
```diff
@@ -54,6 +54,7 @@ identify-demo = "examples.identify.identify:main"
 identify-push-demo = "examples.identify_push.identify_push_demo:run_main"
 identify-push-listener-dialer-demo = "examples.identify_push.identify_push_listener_dialer:main"
 pubsub-demo = "examples.pubsub.pubsub:main"
+kademlia-demo = "examples.kademlia.kademlia:main"
```
Comment (pacrob) [inline pyproject.toml]: Since the kademlia demo is more complicated, I think you can remove it from the `scripts` list here and in `contributing.rst` and just have it be a more manual process.

If you want to update the demo it to use the script syntax instead, that's fine too.

**Comment by pacrob** [comment]
OK, @sumanjeet0012, looking good! A few nits, and I do think having one function that calculates distance would be good. Once these are addressed, I'm good to merge.

**Comment by seetadev** [comment]
@pacrob : Thank you so much Paul for the detailed review and guidance. Appreciate your great support.

We had a discussion on discord and Sumanjeet would address the improvements soon. Thank you once again.

@sumanjeet0012 : Great, thank you for your efforts and initiative. We are finally heading towards having kad-dht in py-libp2p :)

File: libp2p/kad_dht/kad_dht.py (hunk 6 above)
Comment (sumanjeet0012) [reply libp2p/kad_dht/kad_dht.py:371]: In case value not found Added response containing a list of close peers.

File: libp2p/kad_dht/kad_dht.py (hunk 7 above)
Comment (sumanjeet0012) [reply libp2p/kad_dht/kad_dht.py]: Added support for sending a response without a key to indicate PUT_VALUE failure.

File: libp2p/kad_dht/utils.py (hunk 8 above)
Comment (sumanjeet0012) [reply libp2p/kad_dht/utils.py]: refactored the codes, and removed the duplicate function to calculate xor distance.

File: libp2p/kad_dht/kad_dht.py (hunk 9 above)
Comment (sumanjeet0012) [reply libp2p/kad_dht/kad_dht.py]: Added log for both key and value after data storage.

File: libp2p/kad_dht/routing_table.py (hunk 10 above)
Comment (sumanjeet0012) [reply libp2p/kad_dht/routing_table.py]: Removed it.

File: libp2p/kad_dht/routing_table.py (hunk 11 above)
Comment (sumanjeet0012) [reply libp2p/kad_dht/routing_table.py]: Removed it.

File: libp2p/kad_dht/value_store.py (hunk 12 above)
Comment (sumanjeet0012) [reply libp2p/kad_dht/value_store.py]: Corrected.

File: pyproject.toml (hunk 13 above)
Comment (sumanjeet0012) [reply pyproject.toml]: Removed kademlia-demo from both places.

**Comment by sumanjeet0012** [comment]
@seetadev, @pacrob fixed all the highlighted issues.
Ready for review.
